### PR TITLE
Added `WrappedMono` newtype (solved #179)

### PR DESCRIPTION
**Pull request by dcastro**
Ended up not being able to add an instance for `Traversable`/`MonoTraversable` because of the role limitation.

### CODE REVIEW

File: mono-traversable/src/Data/MonoTraversable.hs [hunk 1]
```diff
@@ -1,11 +1,14 @@
-{-# LANGUAGE ConstrainedClassMethods #-}
-{-# LANGUAGE CPP                     #-}
-{-# LANGUAGE DefaultSignatures       #-}
-{-# LANGUAGE FlexibleContexts        #-}
-{-# LANGUAGE FlexibleInstances       #-}
-{-# LANGUAGE TypeFamilies            #-}
-{-# LANGUAGE TypeOperators           #-}
-{-# LANGUAGE UndecidableInstances    #-}
+{-# LANGUAGE ConstrainedClassMethods    #-}
```
Comment (snoyberg) [inline mono-traversable/src/Data/MonoTraversable.hs]: Please undo unrelated whitespace changes.

File: mono-traversable/src/Data/MonoTraversable.hs [hunk 2]
```diff
@@ -1382,3 +1385,25 @@ ointercalate x = mconcat . List.intersperse x . otoList
...
+--
+-- > // package A
+-- > data MyList a = MyList [a] deriving Foldable
+-- >
+-- > // package B
+-- > process :: MonoFoldable mono => mono -> IO ()
+-- >
+-- > // package C
+-- > process (ToMono (MyList []))
+newtype ToMono f a = ToMono { fromMono :: f a }
```
Comment (snoyberg) [inline mono-traversable/src/Data/MonoTraversable.hs]: Looking at prior art, I see [`WrappedMonad`](https://www.stackage.org/haddock/lts-14.7/base-4.12.0.0/Control-Applicative.html#t:WrappedMonad) in base. Perhaps `newtype WrappedMono f a = WrappedMono { unwrapMono :: f a }`?

Comment (dcastro) [reply mono-traversable/src/Data/MonoTraversable.hs]: Agreed. Also changed the wording of the docs to match that of [`WrappedMonoid`](https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Semigroup.html#t:WrappedMonoid)

File: mono-traversable/src/Data/MonoTraversable.hs (hunk 1 above)
Comment (dcastro) [reply mono-traversable/src/Data/MonoTraversable.hs]: Done

File: mono-traversable/src/Data/MonoTraversable.hs [hunk 3]
```diff
@@ -1382,3 +1384,25 @@ ointercalate x = mconcat . List.intersperse x . otoList
...
+-- function that expects a `MonoFoldable`.
+--
+-- > // package A
+-- > data MyList a = MyList [a] deriving Foldable
+-- >
+-- > // package B
+-- > process :: MonoFoldable mono => mono -> IO ()
+-- >
+-- > // package C
+-- > process (WrappedMono (MyList []))
```
Comment (snoyberg) [inline mono-traversable/src/Data/MonoTraversable.hs]: Please also add a minor version bump in the cabal file, a ChangeLog entry, and a `@since` Haddock comment here.

Comment (dcastro) [reply mono-traversable/src/Data/MonoTraversable.hs]: Done

File: mono-traversable/src/Data/MonoTraversable.hs [hunk 4]
```diff
@@ -1382,3 +1384,27 @@ ointercalate x = mconcat . List.intersperse x . otoList
...
+--
+-- @since 1.0.13.0
+newtype WrappedMono f a = WrappedMono { unwrapMono :: f a }
+    deriving newtype (F.Foldable, Functor, Applicative, Monad)
+
+type instance Element (WrappedMono f a) = a
+instance F.Foldable f  => MonoFoldable (WrappedMono f a)
```
Comment (snoyberg) [inline mono-traversable/src/Data/MonoTraversable.hs]: Sorry, I think I reversed things in my brain previously. `WrappedMonad` takes a type with an existing `Monad` instance and creates an `Applicative` instance. This type is doing the opposite: taking something with a polymorphic instance (e.g. `Foldable`) and creating a monomorphic instance (e.g. `MonoFoldable`). I'm now second-guessing my naming choice, and thinking it should be `WrappedPoly`. Though perhaps this type is useful to have as well.

Comment (dcastro) [reply mono-traversable/src/Data/MonoTraversable.hs]: Makes sense, I've renamed `WrappedMono` to `WrappedPoly`!

Regarding a possible `WrappedMono`, I was playing around with that idea and this is what I have so far:

```haskell
{-# LANGUAGE GADTs                   #-}
{-# LANGUAGE InstanceSigs            #-}

data WrappedMono mono a where
    WrappedMono :: Element mono ~ a => mono -> WrappedMono mono a


unwrapMono :: WrappedMono mono a -> mono
unwrapMono (WrappedMono mono) = mono

instance MonoFoldable mono => F.Foldable (WrappedMono mono) where
    foldr :: (a -> b -> b) -> b -> WrappedMono mono a -> b
    foldr f zero (WrappedMono xs) = ofoldr f zero xs
```

First thing I did was tag `WrappedMono mono` with its element type `a`, because foldables have to be of kind `* -> *`.

Second, as far as I can tell, in order for that last line to work, I need to prove that `Element mono ~ a`, so I decided to package up such proof into the `WrappedMono` constructor.
This is not ideal because it means we can't use newtype deriving anymore, but I don't see a way around having to do this.

What do you think?

If the idea seems worth pursuing, I can explore it and add it to this PR or create a separate PR, whichever you prefer.

Comment (snoyberg) [reply mono-traversable/src/Data/MonoTraversable.hs]: Since this one isn't driven by an actual user need, I'm not too concerned about it. If you think it would be useful, we can follow up in a separate PR. I don't want to block your work here on something speculative I just brought up :)

**Comment by snoyberg** [comment]
This PR breaks with GHC 8.0.2. I'm OK dropping support for 8.0.2, but I need to get CI in general updated to allow that to happen. No changes needed from you @dcastro, just explaining why I haven't merged yet.

**Comment by dcastro** [comment]
Thank you :)
